src/i2c.rs

```diff
@@ -0,0 +1,398 @@
+//! Inter Integrated Circuit implementation
+
+use crate::gpio::gpioa::PA8;
+use crate::gpio::gpiob::{PB10, PB11, PB6, PB7, PB8, PB9};
+use crate::gpio::gpioc::PC9;
+use crate::gpio::gpiod::{PD12, PD13};
+use crate::gpio::gpiof::{PF0, PF1, PF14, PF15};
+use crate::gpio::gpioh::{PH11, PH12, PH4, PH5, PH7, PH8};
+use crate::gpio::{Alternate, AF4};
+use crate::rcc::Ccdr;
+use crate::time::Hertz;
+use crate::hal::blocking::i2c::{Write, WriteRead, Read};
+use crate::stm32::{I2C1, I2C2, I2C3, I2C4};
+use cast::{u8, u16};
+
+
+/// I2C error
+#[derive(Debug)]
+pub enum Error {
+    /// Bus error
+    Bus,
+    /// Arbitration loss
+    Arbitration,
+    // Overrun, // slave mode only
+    // Pec, // SMBUS mode only
+    // Timeout, // SMBUS mode only
+    // Alert, // SMBUS mode only
+    #[doc(hidden)]
+    _Extensible,
+}
+
+/// A trait to represent the SCL Pin of an I2C Port
+pub unsafe trait PinScl<I2C> {}
+
+/// A trait to represent the SDL Pin of an I2C Port
+pub unsafe trait PinSda<I2C> {}
+
+pub struct I2c<I2C, PINS> {
+    i2c: I2C,
+    pins: PINS,
+}
+
+macro_rules! busy_wait {
+    ($i2c:expr, $flag:ident) => {
+        loop {
+            let isr = $i2c.isr.read();
+
+            if isr.berr().bit_is_set() {
+                return Err(Error::Bus);
+            } else if isr.arlo().bit_is_set() {
+                return Err(Error::Arbitration);
+            } else if isr.$flag().bit_is_set() {
+                break;
+            } else {
+                // try again
+            }
+        }
+    };
+}
+
+macro_rules! i2c {
+    ($($I2CX:ident: ($i2cX:ident, $i2cXen:ident, $i2cXrst:ident, $apbXenr:ident, $apbXrstr:ident, $pclkX:ident),)+) => {
+        $(
+            impl<SCL, SDA> I2c<$I2CX, (SCL, SDA)> {
+                /// Basically a new function for an I2C peripheral
+                pub fn $i2cX<F> (
+                    i2c: $I2CX,
+                    pins: (SCL, SDA),
+                    freq: F,
+                    ccdr: &Ccdr
+                ) -> Self where
+                    F: Into<Hertz>,
+                    SCL: PinScl<$I2CX>,
+                    SDA: PinSda<$I2CX>,
+                {
+                    ccdr.rb.$apbXenr.modify(|_, w| w.$i2cXen().set_bit());
+                    ccdr.rb.$apbXrstr.modify(|_, w| w.$i2cXrst().clear_bit());
+                    ccdr.rb.$apbXrstr.modify(|_, w| w.$i2cXrst().set_bit());
+
+                    let freq = freq.into().0;
+
+                    assert!(freq <= 1_000_000);
+
+                    let i2cclk = ccdr.clocks.$pclkX().0;    
+
+                    // Refer to figure 539 for this:
```

Comment:
Add document reference

Reply:
How exactly do I add a document reference? (Never done that before)

Reply:
Ah, I just mean reference the ST documentation number and revision in the comment.
Something like `// RM0433 Rev 6 - Figure 539`